### PR TITLE
handle if dot is not present

### PR DESCRIPTION
**Pull request by kshitij12345**
Native grahpviz installation is required for `torchviz/pygraphviz` to work. If it is not found, we warn the users.

Also, we update code to save visualization output to unique files and clean-up the file post upload.

Ref: https://github.com/pydot/pydot/issues/257

### CODE REVIEW

File: src/neptune_pytorch/impl/__init__.py [hunk 1]
```diff
@@ -42,6 +42,7 @@
 IS_TORCHVIZ_AVAILABLE = True
 try:
     import torchviz
+    from graphviz import ExecutableNotFound
```
Comment (kshitij12345) [inline src/neptune_pytorch/impl/__init__.py:45]: torchviz depends on graphviz

File: src/neptune_pytorch/impl/__init__.py [hunk 2]
```diff
@@ -154,9 +155,13 @@ def hook(module, input, output):
                 dot = torchviz.make_dot(output, params=dict(module.named_parameters()))
                 # Use tempfile correctly.
                 dot.format = "png"
-                dot.render(outfile="torch-viz.png")
-                self._namespace_handler["model"]["visualization"].upload("torch-viz.png")
-                self._is_viz_saved = True
+                try:
+                    dot.render(outfile="torch-viz.png")
+                    self._namespace_handler["model"]["visualization"].upload("torch-viz.png")
+                except ExecutableNotFound:
```
Comment (kshitij12345) [inline src/neptune_pytorch/impl/__init__.py:161]: If native package for rendering `dot` is not installed then `graphviz` throws this error.

File: src/neptune_pytorch/impl/__init__.py [hunk 3]
```diff
@@ -205,6 +210,10 @@ def __del__(self):
 
         if self._vis_hook_handler is not None:
             self._vis_hook_handler.remove()
+            fn = "torch-viz.gv"
```
Comment (kshitij12345) [inline src/neptune_pytorch/impl/__init__.py:213]: clean-up

File: src/neptune_pytorch/impl/__init__.py [hunk 4]
```diff
@@ -152,11 +154,19 @@ def add_visualization_hook(self):
...
+                # generate unique name so that multiple concurrent runs
+                # don't over-write each other.
+                viz_name = str(uuid.uuid4()) + ".png"
+                try:
+                    dot.render(outfile=viz_name)
+                    safe_upload_visualization(self._namespace_handler["model"], "visualization", viz_name)
+                except ExecutableNotFound:
+                    # This errors because `dot` renderer is not found even
+                    # if python binding of `graphviz` are available.
+                    warnings.warn("Skipping model visualization because no dot (graphviz) installation was found.")
```
Comment (kshitij12345) [inline src/neptune_pytorch/impl/__init__.py:167]: cc: @normandy7 for the message.

Comment (normandy7) [reply src/neptune_pytorch/impl/__init__.py:167]: Looks good.

**Review by normandy7 — APPROVED** [review]

File: src/neptune_pytorch/impl/__init__.py [hunk 5]
```diff
@@ -152,11 +154,19 @@ def add_visualization_hook(self):
...
+                finally:
+                    self._is_viz_saved = True
```
Comment (AleksanderWWW) [inline src/neptune_pytorch/impl/__init__.py:169]: Not sure about this part - `finally` will execute no matter what. If it's just about setting a variable, does it need to be in this block?

File: src/neptune_pytorch/impl/__init__.py [hunk 6]
```diff
@@ -207,7 +217,23 @@ def __del__(self):
             self._vis_hook_handler.remove()
 
 
-def safe_upload(run, name, model):
+def safe_upload_visualization(run, name, file_name):
```
Comment (AleksanderWWW) [inline src/neptune_pytorch/impl/__init__.py:220]: One additional idea: would it be possible to add type hints to the function and method params?

Comment (AleksanderWWW) [reply src/neptune_pytorch/impl/__init__.py:220]: Can be also a separate PR, but at some point would be nice to have them - it's then much easier to maintain :smile:

File: src/neptune_pytorch/impl/__init__.py (hunk 5 above)
Comment (kshitij12345) [reply src/neptune_pytorch/impl/__init__.py]: Removed. Thanks!

**Review by AleksanderWWW — APPROVED** [review]

**Pull request merged**